FixMath2.h

```diff
@@ -0,0 +1,845 @@
+ *
+ */
+
+
+/** This file implements two fixed point number classes. These numbers can have a fractional
+part but are actually standard integers under the hood which makes calculations with them
+efficient on platforms which do not have a FPU like most micro-controllers. These numbers can be
+signed (SFixMath2) or unsigned (UFixMath2).
+
+A fixed point number has its range defined by the number of bits encoding the integer part (NI 
+in the following) and its precision by the number of bits encoding the fractional part (NF).
+
+Like standard C(++) types, the fixed point numbers defined here are following some rules:
+ - any fixed type can be converted to another *as long as the value can be represented in the destination type*. Casting to a bigger type in term of NI and NF is safe, but reducing NI can lead to an overflow if the new type cannot hold the integer value and reducing NF leads to a loss of precision.
+ - all operations between fixed point number is safe (it won't overflow) and preserve the precision. In particular:
+   - only addition, subtraction and multiplication are implemented
+   - any operation between a signed and an unsigned leads to a signed number
+   - resulting numbers will be casted to a type big enough to store the expected values. It follows that it is worth starting with types that are as small as possible to hold the initial value.
+ - all operations between a fixed point number and a native type (int, float, uint) are *not* safe. If the resulting value cannot be represented in the fixed point type it will overflow. Only addition, subtraction, multiplication and right/left shift are implemented.
+ - safe right/left shifts, which return the correct value in the correct type are implemented as .sR<shift>() and .sL<shift>() respectively, shift being the shifting amount.
+*/
+
+
+
+
+#ifndef FIXMATH2_H_
+#define FIXMATH2_H_
+
+#include "IntegerType.h"
+
+#define SHIFTR(x,bits) (bits > 0 ? (x >> (bits)) : (x << (-bits))) // shift right for positive shift numbers, and left for negative ones.
+#define MAX(N1,N2) (N1 > N2 ? N1 : N2)
+
+// Experiments
+/*#define NBITSREAL(X,N) (abs(X) < (1<<N) ? N : NBITSREAL2(X,N+1))
+  #define NBITSREAL2(X,N) (abs(X) < (1<<N) ? N : NBITSREAL(X,N+1))
+  #define UFixAuto(X) (UFixMath2<NBITSREAL(X,0),0>(X))
+  #define UFixAuto(X) (UFixMath2<NBITSREAL(X,0),0>(X))*/
+
+/*
+  template<typename T>
+  constexpr byte nBitsReal(T x, byte n=0) {
+  if (abs(x) < ((T)1 << n)) {
+  return n;
+  } else {
+  return nBitsReal(x, n + 1);
+  }
+  }
+
+ #define UFixAuto(X) (UFixMath2<nBitsReal(X),0>(X))
+*/
+
+ 
+
+
+
+
+// Forward declaration
+template<byte NI, byte NF>
+class SFixMath2;
+
+
+
+/** Instanciate an unsigned fixed point math number.
+@param NI The number of bits encoding the integer part
+@param NF The number of bits encoding the fractional part
+*/
+template<byte NI, byte NF=0> // NI and NF being the number of bits for the integral and the fractionnal parts respectively.
```

Comment:
Default to 8 for NF? 0 seems a bit pointless.

Reply:
That's a remain of trying to make an automatic constructor for integers which would find the smallest working `NI` and putting `NF=0`. Unfortunately I did not manage to make that work (too bad macros cannot be recursive). So I think I'll remove this default value as a whole, it is more asking for wrong usage than anything else.